### ISSUE TITLE
[Feature request] Allow reading gzipped casts

### ISSUE THREAD

**Issue by itskenny0** [asciinema/asciinema-player#137]
I would like to store the casts on our own (custom) asciinema server in gzipped format, however asciinema-player does not support reading gzipped content.

**Comment by RobertLarsen** [comment]
I would very much like this feature too...or bzip2 which in my case compresses twice as well as gzip.

**Comment by mathieu-aubin** [comment]
You could store them as gzip and use nginx static gzip module to send them, maybe? i suggest you have a look at it

**Comment by mathieu-aubin** [comment]
You could store them as gzip and use [nginx gzip static](http://nginx.org/en/docs/http/ngx_http_gzip_static_module.html) module to send them, maybe? i suggest you have a look at it

**Comment by krader1961** [comment]
Being able to handle gzip'd data would be extremely useful as it would allow using a `data` URI. I'm working on switching the Elvish shell documentation from using short tty snapshots created with a custom, hard to use, mechanism to an Asciinema based solution. Gzip'ing the resulting asciicasts makes them approximately eight times smaller. Small enough that even with base64 encoding they could still be embedded using a `data:gzip;base64,...` URI. Thus avoiding the need for additional HTTP requests.

**Comment by sickill** [comment]
That sounds pretty useful indeed. 

I quickly searched for built-in decompression support in JS and found this: https://developer.mozilla.org/en-US/docs/Web/API/DecompressionStream

I'd prefer to use a browser API like this, however it looks like FF and Safari don't support this yet. It seems to be a standard draft: https://wicg.github.io/compression/

Assuming this would eventually get adopted by most browsers we could use it at some point.

**Comment by mathieu-aubin** [comment]
it does - the reason i mentionned using web server gzip/unzip feature natively was that i am of opinion that it is very imprudent to run stuff withiut proxying thru a dedicated, well supported and secure web server. Otherwise, i agree the feature could be useful if no extra commands/functions ajd pipelines is desired.

**Comment by sickill** [comment]
I've just released v3.0.0 which allows using custom loader function, which could for example be used to load gzipped/base64ed inline asciicasts like this:

```javascript
const asciicast = 'H4sIAAAAAAACA+2YzW7bRhCA732KtXrpIZH2/8ds7TaGATVOdGiLXGwXoOSVyZZLKhSlxA2SQy+9JQ/Ta29tXqRP0qVEcilKFOUiKHyoYBi75Oy388OdGfJNb6nTeZjEvWOAH4Heq/AmC+wYQWRngQ5vgyy/Re0sC42eZ76Z5fc5Q4QoxYW9oeOlvfSm9/3w/NkzO+oNFvN0MA7jwTScBz0r8cP5d8/zG68znZrHmPFJEiVp7+3bzy5hH2IsJbdSSS56tYAQjS9PMYQ06F2vBAghVGwKXEPvXT6ColhAoFmPvnhSXDFXaTFS2Bg/nYRxU+Lr5oXZwmrYvAjKLbB5t7VHY35SCl9UuyN+trZDCYQ4Ku34+IeTEJUEZtBZOi43lh72MGMe9KDZXNdufU5EfYyxELTh3A3en787nuzkUY4Rwvt4H391PNXJswZjyvbx/vqt4mHYwcP2WRGU7MMBR0MH0KQNiYvHxuK1CCMSySpkMyeAKwGFSUtMoYcEy1Ur9r+kZoPQpR+jmAq1y1zJ1n8mTX7Sk2w+uA+WC0zlTi/uUNjfcwCbRwGTs3IHgiGvOaXFLTO/vrZLb6GYDcZ+vWMHpJ1ASRFjYj8wdEDWCVQcc9GhoXZA3gEkfSwRZHQ/8MgBRSeQSutDcTBQdgI5RVigg4FdKYP2EYdNg4X58WxvHYhLvPn7w/ut/P4f/K5Su7FLHxcPqTLRPIkpUR3H5a4Cs5FUQH+SxNPwdhCSwXp0WE1qHnNlw4+YWX7STdYmcSmgbJ60iw2Ceyzac9eqD4lKpOJquw9ZggM6kTVBQKi2zsIpglQFVXbOq1RWm2WVWCHzVXFhWAo9LVRByDvd0GTdRBWTE+phZxULCo2QwohuNV8savFHc1Pr0tKJipYP4crP8f+Tq+axthnPE+ZylICRb/T1p0sv8BF8jNz0mygqM6NH+bAZy9GLb5+DJeyLPixj52E1HC1DA8I5SGY6BvNkkU408OMbME21ju7ATTjP0nC8yPxxpMtjBD0Kh0GWzebHg0GsE4voh8ng80nglw8uss2XHGZ3Mw3AcaCjGYitlDvy9rAj8+V5bPPbSUtqDqfgLlkAP9Ug1q+OykyGSI08CfTk50D7URbcg50l1lz7ZhP+osHIqYVojfzyXrqWP0vWr8PMVYCCzBreuD98mqRg5ccGWXhEDJ/oSWJs3ECqb23AdKpvwAsb18Vcp0dVn9+MSCl8sDK5DmFs/xs/s6+PoEpBeCMDYfCy606e/yAOahMeNJKNS1cSUonb09VqeZXxymWsT6miRLYvw9RTaHjcjFV9/x2LSOmQp8fdRla6KIQV26FKU5JbrQUWO411LayHhtvrGGeMtVrboeW95PPQ1X3Po9Y66sioEDqpF0Dn1Y0CSGoFcFWK64GOXGmDZtsNXHLVdN+DaQALHQXhgrV++7ACEmKu/sW3jwfUX4o+4Ywr90J5SOPV0tjlLA7Jzs9FeZv2D9AGNmLREgAA'

function decodeAndDecompress(gzippedBase64Data) {
  const gzippedData = atob(gzippedBase64Data);
  const gzippedDataArray = Uint8Array.from(gzippedData, c => c.charCodeAt(0));
  const ds = new DecompressionStream('gzip');
  const writer = ds.writable.getWriter();
  writer.write(gzippedDataArray);
  writer.close();

  return new Response(ds.readable).arrayBuffer().then(function (arrayBuffer) {
    return new TextDecoder().decode(arrayBuffer);
  });
}

AsciinemaPlayer.create(
  { data: () => decodeAndDecompress(asciicast) },
  document.getElementById('demo')
);
```

As I mentioned earlier the compression APIs are not well supported yet so the above example works in Chrome only at the moment. Disclaimer: `decodeAndDecompress` function is something I stitched together from various bits I found on the internet, and I'm not sure if it's the most optimal way to combine base64 decoding with gzip decompression. But at least it works :)

**Comment by sickill** [comment]
I'd reconsider built-in support for `data:gzip;base64,...` when most browsers start supporting the above API. Until then feel free to use player v3.0 with a snippet like above.
